core/pta/attestation.c

```diff
@@ -0,0 +1,761 @@
+	if (!res)
+		DMSG("Loaded %zu bit key pair", crypto_bignum_num_bits(key->n));
+
+out:
+	free(buf);
+	return res;
+}
+
+static TEE_Result write_key(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *buf = NULL;
+	size_t size = 0;
+
+	DMSG("Saving key pair");
+	res = serialize_key(&buf, &size);
+	if (res)
+		return res;
+
+	res = sec_storage_obj_write(&pta_uuid, TEE_STORAGE_PRIVATE,
+				    key_file_name, sizeof(key_file_name) - 1,
+				    buf, size, 0, TEE_DATA_FLAG_ACCESS_WRITE);
+	if (!res)
+		DMSG("Wrote %zu bytes", size);
+	free(buf);
+	return res;
+}
+
+static TEE_Result init_key(void)
+{
+	TEE_Result res = TEE_SUCCESS;
+
+	if (!key) {
+		res = load_key();
+		if (res) {
+			res = generate_key();
+			if (res)
+				return res;
+			res = write_key();
+			if (res)
+				return res;
+		}
+	}
+	return res;
+}
+
+static TEE_Result cmd_get_pubkey(uint32_t param_types,
+				 TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *e = params[0].memref.buffer;
+	uint32_t *e_out_sz = &params[0].memref.size;
+	uint8_t *n = params[1].memref.buffer;
+	uint32_t *n_out_sz = &params[1].memref.size;
+	size_t sz = 0;
+
+	if (param_types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					   TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					   TEE_PARAM_TYPE_NONE,
+					   TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = init_key();
+	if (res)
+		return res;
+
+	sz = crypto_bignum_num_bytes(key->e);
+	if (*e_out_sz >= sz)
+		crypto_bignum_bn2bin(key->e, e);
+	else
+		res = TEE_ERROR_SHORT_BUFFER;
+	*e_out_sz = sz;
+
+	sz = crypto_bignum_num_bytes(key->n);
+	if (*n_out_sz >= sz)
+		crypto_bignum_bn2bin(key->n, n);
+	else
+		res = TEE_ERROR_SHORT_BUFFER;
+	*n_out_sz = sz;
+
+	return res;
+}
+
+static TEE_Result hash_binary(const TEE_UUID *uuid, uint8_t *hash)
+{
+	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
+	unsigned int tag_len = FILE_TAG_SIZE;
+	const struct ts_store_ops *ops = NULL;
+	struct ts_store_handle *h = NULL;
+
+	SCATTERED_ARRAY_FOREACH(ops, ta_stores, struct ts_store_ops) {
+		res = ops->open(uuid, &h);
+		if (!res)
+			break;  /* TA found */
+	}
+	if (res)
+		return res;
+
+	/*
+	 * Output hash size is assumed to be the same size as the file tag
+	 * size which is the size of the digest in the TA shdr. If one or the
+	 * other changes, additional hashing will be needed.
+	 */
+	COMPILE_TIME_ASSERT(FILE_TAG_SIZE == TEE_SHA256_HASH_SIZE);
+	assert(ops);
+	res = ops->get_tag(h, hash, &tag_len);
+	if (res)
+		goto out;
+
+	DMSG("TA %pUl hash:", uuid);
+	DHEXDUMP(hash, TEE_SHA256_HASH_SIZE);
+out:
+	ops->close(h);
+	return res;
+}
+
+/* Hash @nonce and @hash into @digest */
+static TEE_Result digest_nonce_and_hash(uint8_t *digest, uint8_t *nonce,
+					size_t nonce_sz, uint8_t *hash)
+{
+	TEE_Result res = TEE_SUCCESS;
+	void *ctx = NULL;
+
+	res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SHA256);
+	if (res)
+		return res;
+
+	res = crypto_hash_init(ctx);
+	if (res)
+		goto out;
+	if (nonce) {
```

Comment:
No, but it would make the documentation more complex :wink "SIG(SHA256(Nonce | digest)) except when no Nonce is provided in which case it is SIG(digest)". Perhaps I should simply reject `nonce == NULL`? It doesn't seem safe to run the algorithm without a nonce.

Reply:
I agree, we should reject to perform a signature without nonce.